### ISSUE TITLE
Command timeout with Arista vEOS

### ISSUE THREAD

**Issue by aecro** [trigger/trigger#238]
Hey,

I'm new to trigger and want to test it first before using it with production devices, so I set up a couple of Arista vEOS VMs. Connecting to the VM using dev.connect() works just fine. However, if I try to use run_cmds or the Commando class I get timeouts like this:

```
$ run_cmds -d 10.40.55.10 -c "show clock" --push
10.40.55.10 - Error: Timed out while sending commands
```

```
$ python test.py 
Unhandled Error
Traceback (most recent call last):
Failure: trigger.exceptions.CommandTimeout: Timed out while sending commands

{}
```
## test.py
```
from trigger.cmds import Commando

if __name__ == '__main__':
    device_list = ['10.40.55.10']
    command_list = ['show clock']
    showclock = Commando(devices=device_list,commands=command_list)
    showclock.run()
    print showclock.results
```

**Comment by aecro** [comment]
I have a feeling this is related to auto-enable not working with vEOS (all Aristas?). I tried both ways of executing "show clock" like above with a Brocade switch and they worked. 
Also, when using gong to connect to that Brocade switch I get auto-enabled, where with connecting to my vEOS VM I need to manually type enable and manually enter the enablePW, even when it's specified in the netdevices.json.
On a side note: both Arista and Brocade allow "show clock" outside enable mode, so an enable password should not be necessary.

**Comment by jathanism** [comment]
Hi @aecro.

This may very well be related to the prompt pattern matching on the login phase. Can you please provide debug logs for when you connect to your vEOS device? (Please be sure to "sanitize" any private data if necessary)

You may do this by prepending your command with `DEBUG=1` (or by setting the `DEBUG` environment variable):

```bash
$ DEBUG=1 run_cmds -d 10.40.55.10 -c "show clock" --push
```

You'll see a lot of logs spew to the console. Paste those in a Gist and link them here, please! :8ball:

**Comment by aecro** [comment]
Hey again,

apologies for the delay, holidays and stuff.
Here is the requested output: https://gist.github.com/anonymous/217c2ce575e4c7b4c804

**Comment by jathanism** [comment]
Oh snap! I'll take a look soon.

**Comment by jathanism** [comment]
Ok, so now it's my turn to apology for the delay: holidays and stuff and coming back to work sick!

I'm realizing there a couple issues you're running into. Let's start with the auto-enabling issue.

First things first: I've been able to identify this as a bug.

I have a vEOS VM, and set my user to priv-lvl 1 so that enable is required. The connection is hanging at the `Password: ` prompt. Here's what I'm seeing in my logs if I set an enable password (of `password`):

```
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] BYTES: 'arista-vm>'
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] Enable prompt detected: '\x1b[5narista-vm>'
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] Enable required, sending enable commands
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] Enable password detected, sending...
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] BYTES: 'e'
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] BYTES: 'n'
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] BYTES: 'a'
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] BYTES: 'b'
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] BYTES: 'l'
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] BYTES: 'e'
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] BYTES: '\r\n'
2016-01-12 10:42:51-0800 [SSHChannel session (0) on SSHService ssh-connection on TriggerSSHTransport,client] [arista-sw1] BYTES: 'Password: '```

Observe it's trying to send the enable password BEFORE the `Password: ` prompt even appears. BUG!!

I have a feeling this is related to a change that was made to alleviate another bug in Trigger 1.5.x. I'm going to poke around a bit and post back.

**Comment by jathanism** [comment]
@aecro In the interim, here is a temporary workaround that should disable the auto-enable behavior for you, because said `show clock` should be allowed outside of enable mode, and you're right!

Unfortunately the assumption right now is that if a `>` (non-enabled) prompt is detected on an IOS-like (any device platform that emulates IOS like Brocade, Foundry, Arista, etc.), auto-enable is automatic.

There's not currently a way to toggle this behavior, because this feature was implemented with the assumption that you'll always want to enable.

There is, however, a way to change the behavior of Trigger into not requiring enable at runtime using two undocumented settings. 

### Prompt settings

`IOSLIKE_ENABLE_PAT` which is a regex pattern for what Trigger expects a non-enable prompt to look like, which is effectively looking for `> `:

```
>>> from trigger.conf import settings
>>> print settings.IOSLIKE_ENABLE_PAT
\S+(\(config(-[a-z:1-9]+)?\))?>\s?$
```

`IOSLIKE_PROMPT_PAT`, which is used to determine if you're already enabled, and is effectively looking for `# ` :

```
>>> print settings.IOSLIKE_PROMPT_PAT
\S+(\(config(-[a-z:1-9]+)?\))?#\s?$
```

Summary:

- `IOSLIKE_ENABLE_PAT`: Regex pattern used to detect if we NEED to enable
- `IOSLIKE_PROMPT_PAT`: Regex pattern used to detect if we are already enabled and may proceed w/ state machine

### Workaround1: Enable prompt detection patterns

The workaround is simple. 

1. Replace `IOSLIKE_ENABLE_PAT` with a bogus pattern that will never match, so that auto-enable is never required
2. Replace `IOSLIKE_PROMPT_PAT` with the original value of `IOSLIKE_ENABLE_PAT`, so that `>` is the desired state.
3. Profit!

So here's what I put in my `settings.py`:

```python
# This is our NON-ENABLED prompt (used to detect if we NEED to enable)
# IOSLIKE_ENABLE_PAT = r'\S+(\(config(-[a-z:1-9]+)?\))?>\s?$'
IOSLIKE_ENABLE_PAT = r'bogus$'  # This should never match!

# This is our desired operational prompt
# IOSLIKE_PROMPT_PAT = r'\S+(\(config(-[a-z:1-9]+)?\))?#\s?$'
IOSLIKE_PROMPT_PAT = r'\S+(\(config(-[a-z:1-9]+)?\))?>\s?$'  # It's now expecting '>'
```

And here's the proof that `run_cmds` now works:

```
$ run_cmds -d arista-sw1 -c 'show clock' --push

###
# arista-sw1
###
arista-sw1# show clock
Sun Jun 14 13:17:38 2015
Timezone: UTC
Clock source: local

Done.
```

**NB:** I realize this isn't practical in a multi-vendor environment, but I wanted to give you the background on how enable prompt detection works for IOS-like devices.

### Workaround 2: Skipping IOS-like handling
 
Now that you understand the prompt patterns (I hope), this workaround is a lot simpler and is vendor-specific.

There are two settings at play here, one of which which is documented:

- [`IOSLIKE_VENDORS`](http://trigger.readthedocs.org/en/latest/usage/configuration.html#ioslike-vendors): A list of vendor names that should be considered IOS-like.
- `PROMPT_PATTERNS`: A mapping of vendor:regex of prompt patterns to use

First, we want to tell Trigger not to treat Arista as IOS-like, so remove `arista` from the list:

```python
# IOSLIKE_VENDORS = ['a10', 'arista', 'brocade', 'cisco', 'dell', 'foundry']  # Default
IOSLIKE_VENDORS = ['a10', 'brocade', 'cisco', 'dell', 'foundry']
```

Next, we want to add an entry for `arista` in `PROMPT_PATTERNS`. We'll use the NON-enabled regex value from `IOSLIKE_ENABLE_PAT` for this:

```python
# Mapping of per-vendor prompt patterns
PROMPT_PATTERNS = {
    'arista': r'\S+(\(config(-[a-z:1-9]+)?\))?>\s?$'  # Same value as IOSLIKE_ENABLE_PAT
    # other vendor patterns ...
}
```

And now you have a vendor-specific workaround for Arista devices that don't require enable. If you execute `run_cmds` again, it works:

```
$ run_cmds -d arista-sw1 -c 'show clock' --push                                                                                                                              <<<

###
# arista-sw1
###
arista-sw1# show clock
Sun Jun 14 13:34:37 2015
Timezone: UTC
Clock source: local

Done.
```

(I really need to update the clock in my VM!)

Finally... I'll go back to digging into that bug.

**Comment by jathanism** [comment]
Fixed in 575939f6312b37c60785f10f8a7dddd384b058d9
